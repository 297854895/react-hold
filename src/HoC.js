import React, { Component } from 'react'
import { findDOMNode } from 'react-dom'
import {
  isNull, isObject, isFunction,
  getNodeSize, getComputedStyle, getComponentName,
  addHandler, removeHandler,
} from './utils'
import Fill from './holders/Fill'
import createRefiter from './createRefiter'

const $nbsp = '\u00A0'

const window = global

const envStyle = {
  position: 'relative',
  padding: '0px',
  margin: '0px',
  width: '100%',
  height: '100%',
  border: 'none',
  overflow: 'visible',
}

/**
 * Hold the target component,
 * returns a holdable higher-order component.
 *
 * @param {Component} targetComponent
 * @param {Function} condition
 * @param {Component} holder
 * @param {Object} holderProps
 * @returns {Component}
 */
export default function (targetComponent, condition, holder = Fill, holderProps = {}) {
  if (!isFunction(targetComponent) && typeof targetComponent !== 'string') {
    throw new TypeError('Expected the target component to be a react or dom component.')
  }

  if (!isFunction(condition)) {
    throw new TypeError('Expected the hold condition to be a function.')
  }

  if (isObject(holder)) {
    holderProps = holder
    holder = Fill
  }

  holderProps.color = holderProps.color || '#eee'
  holderProps.width = !isNull(holderProps.width) ? holderProps.width : null
  holderProps.height = !isNull(holderProps.height) ? holderProps.height : null

  const wrappedComponentName = getComponentName(targetComponent)

  const refiter = createRefiter(targetComponent)

  class Hold extends Component {

    constructor(...args) {
      super(...args)

      this.state = {
        hold: true,
        copy: true,
        color: holderProps.color, // holder's color
        width: holderProps.width, // holder's width
        height: holderProps.height, // holder's height
      }
      // The style value of original node
      this.originNodeStyle = null
      this.resizeHandler = this.updateHolderSizeIfNecessary.bind(this)
    }

    componentWillMount() {
      if (condition.call(null, this.props)) {
        refiter.refit()
      } else {
        this.cancelHold()
      }
    }

    componentDidMount() {
      this.originNodeStyle = this.computeOriginNodeStyle()
      this.setState({ copy: false })

      addHandler(window, 'resize', this.resizeHandler)
    }

    componentWillReceiveProps(nextProps) {
      if (condition.call(null, nextProps)) {
        this.setState({
          hold: true,
          copy: true,
        })
      } else {
        this.cancelHold()
      }
    }

    componentDidUpdate() {
      if (this.state.hold) {
        if (this.state.copy) {
          refiter.refit()
          this.originNodeStyle = this.computeOriginNodeStyle()
          this.setState({ copy: false })
        } else if (!isNull(this.originNodeStyle)) {
          this.setFakeNodeStyle(this.originNodeStyle)
          this.updateHolderSizeIfNecessary()
          this.originNodeStyle = null
        }
      }
    }

    componentWillUnmount() {
      removeHandler(window, 'resize', this.resizeHandler)
    }

    setFakeNodeStyle(style) {
      if (!isObject(style)) return

      const fake = this.refs.fake
      // hidden element
      fake.style.display = 'none'
      // set style
      Object.keys(style).forEach((name) => {
        if (name !== 'display') {
          fake.style[name] = style[name]
        }
      })
      // fix style
      fake.style.opacity = 1
      fake.style.background = 'transparent'
      fake.style.borderColor = 'transparent'
      let display = style.display
      if (display === 'inline') display = 'inline-block'
      fake.style.display = display
    }

    computeOriginNodeStyle() {
      let result = null
      const originNode = findDOMNode(this)

      // store original display property
      let computedStyle = getComputedStyle(originNode, null)
      const originDisplay = computedStyle.getPropertyValue('display')

      // set display to 'none' before recompute is very **important**,
      // don't remove or move this step!
      originNode.style.display = 'none'
      // compute node style
      computedStyle = getComputedStyle(originNode, null)

      Object.keys(computedStyle).forEach((key) => {
        if (/[0-9]+/.test(key)) {
          const name = computedStyle[key]
          result = result || {}
          if (name === 'display') {
            result[name] = originDisplay
          } else {
            result[name] = computedStyle.getPropertyValue(name)
          }
        }
      })

      return result
    }

    cancelHold() {
      refiter.undo()
      this.setState({ hold: false })
    }

    updateHolderSizeIfNecessary() {
      const { env } = this.refs
      if (!env) return

      const size = getNodeSize(env)
      this.setState({
        width: isNull(holderProps.width) ? size.width : holderProps.width,
        height: isNull(holderProps.height) ? size.height : holderProps.height,
      })
    }

    render() {
      const { hold, copy, color, width, height } = this.state

      if (!hold || copy) {
        const { innerRef, ...propsForElement } = this.props
        if (innerRef && !hold) propsForElement.ref = innerRef
        return React.createElement(targetComponent, propsForElement)
      }

      const propsForHolder = { ...holderProps, color, width, height }
      if (typeof propsForHolder.children === 'string') {
        propsForHolder.children = propsForHolder.children.replace(/ /g, $nbsp)
      }
      propsForHolder.children = propsForHolder.children || $nbsp.repeat(4)

      return (
        <div ref="fake">
          <div ref="env" style={envStyle}>
            { React.createElement(holder, propsForHolder) }
          </div>
        </div>
      )
    }
  }

  Hold.displayName = `Hold(${wrappedComponentName})`

  return Hold
}
