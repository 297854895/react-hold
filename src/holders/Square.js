import React from 'react'
import PropTypes from 'prop-types'
import shapes from '../shapes'
import { CENTER } from '../align'
import { isNull } from '../utils'

const Square = ({ color, width, height, children, side, align }) => {
  if (isNull(side)) {
    if (!isNull(width) && !isNull(height)) {
      side = width > height ? height : width
    } else if (!isNull(width)) {
      side = width
    } else if (!isNull(height)) {
      side = height
    }
  }
  const lineHeight = (typeof side === 'string' && side.trim()) ?
    side : typeof side === 'number' ?
    `${side}px` : null

  return (
    <div style={{ textAlign: align }}>
      <div
        style={{
          display: 'inline-block',
          textAlign: 'center',
          background: color,
          width: side,
          height: side,
          lineHeight,
        }}
      >
        { children }
      </div>
    </div>
  )
}

Square.propTypes = {
  ...shapes,
  side: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  align: PropTypes.string,
}

Square.defaultProps = {
  width: null,
  height: null,
  side: null,
  align: CENTER,
}

export default Square
