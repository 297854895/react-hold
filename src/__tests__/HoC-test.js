import React from 'react'
import PropTypes from 'prop-types'
import { mount } from 'enzyme'
import HoC from '../HoC'
import Square from '../holders/Square'

const P = HoC('p', props => !props.children)

describe('Holdable HoC component', () => {
  it('Create correctly', () => {
    expect(P).toHaveProperty('displayName', 'Hold(p)')
    expect(P.propTypes.holder).not.toBeNull()
    expect(P.propTypes.holderProps).not.toBeNull()
    expect(P.propTypes.props).not.toBeNull()
    expect(P.propTypes.innerRef).not.toBeNull()
    expect(P.defaultProps.holder).not.toBeUndefined()
    expect(P.defaultProps.holderProps).not.toBeUndefined()
    expect(P.defaultProps.props).not.toBeUndefined()
    expect(P.defaultProps.innerRef).not.toBeUndefined()
  })

  it('Do not hold', () => {
    const wrapper = mount(<Container presentation={P}>foo</Container>)
    expect(wrapper).toMatchSnapshot()
  })

  it('Hold', () => {
    const wrapper = mount(<Container presentation={P} />)
    expect(wrapper).toMatchSnapshot()

    wrapper.setState({ children: 'I am child, then cancel hold.' })
    expect(wrapper).toMatchSnapshot()

    wrapper.setState({
      props: {
        name: 'bar',
        holder: Square,
        holderProps: {
          color: 'green',
          width: 40,
          height: 30,
        },
      },
      children: '',
    })
    expect(wrapper).toMatchSnapshot()
  })
})

class Container extends React.Component {
  constructor(...args) {
    super(...args)
    this.state = {
      props: {
        name: 'foo',
        holder: undefined,
        holderProps: null,
      },
      children: this.props.children,
    }
  }

  render() {
    const { presentation } = this.props
    const { props, children } = this.state
    return (
      <div>
        { React.createElement(presentation, props, children) }
      </div>
    )
  }
}

Container.propTypes = {
  presentation: PropTypes.func.isRequired,
  children: PropTypes.string,
}

Container.defaultProps = {
  children: '',
}
