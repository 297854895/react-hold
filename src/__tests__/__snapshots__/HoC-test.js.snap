// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Holdable HoC component Do not hold 1`] = `
<Container
  presentation={[Function]}
>
  <div>
    <Hold(p)
      holder={[Function]}
      holderProps={null}
      innerRef={null}
      name="foo"
      props={null}
    >
      <p
        name="foo"
      >
        foo
      </p>
    </Hold(p)>
  </div>
</Container>
`;

exports[`Holdable HoC component Hold 1`] = `
<Container
  presentation={[Function]}
>
  <div>
    <Hold(p)
      holder={[Function]}
      holderProps={null}
      innerRef={null}
      name="foo"
      props={null}
    >
      <div>
        <div
          style={
            Object {
              "border": "none",
              "height": "100%",
              "margin": "0px",
              "overflow": "visible",
              "padding": "0px",
              "position": "relative",
              "width": "100%",
            }
          }
        >
          <Fill
            align="center"
            cancelHold={[Function]}
            color="#eee"
            height={0}
            targetProps={
              Object {
                "children": "",
                "name": "foo",
              }
            }
            width={0}
          >
            <div
              style={
                Object {
                  "textAlign": "center",
                }
              }
            >
              <div
                style={
                  Object {
                    "background": "#eee",
                    "display": "inline-block",
                    "height": 0,
                    "lineHeight": "0px",
                    "textAlign": "center",
                    "width": 0,
                  }
                }
              >
                          
              </div>
            </div>
          </Fill>
        </div>
      </div>
    </Hold(p)>
  </div>
</Container>
`;

exports[`Holdable HoC component Hold 2`] = `
<Container
  presentation={[Function]}
>
  <div>
    <Hold(p)
      holder={[Function]}
      holderProps={null}
      innerRef={null}
      name="foo"
      props={null}
    >
      <p
        name="foo"
      >
        I am child, then cancel hold.
      </p>
    </Hold(p)>
  </div>
</Container>
`;

exports[`Holdable HoC component Hold 3`] = `
<Container
  presentation={[Function]}
>
  <div>
    <Hold(p)
      holder={[Function]}
      holderProps={
        Object {
          "color": "green",
          "height": 30,
          "width": 40,
        }
      }
      innerRef={null}
      name="bar"
      props={null}
    >
      <div>
        <div
          style={
            Object {
              "border": "none",
              "height": "100%",
              "margin": "0px",
              "overflow": "visible",
              "padding": "0px",
              "position": "relative",
              "width": "100%",
            }
          }
        >
          <Square
            align="center"
            cancelHold={[Function]}
            color="green"
            height={30}
            side={null}
            targetProps={
              Object {
                "children": "",
                "name": "bar",
              }
            }
            width={40}
          >
            <div
              style={
                Object {
                  "textAlign": "center",
                }
              }
            >
              <div
                style={
                  Object {
                    "background": "green",
                    "display": "inline-block",
                    "height": 30,
                    "lineHeight": "30px",
                    "textAlign": "center",
                    "width": 30,
                  }
                }
              >
                          
              </div>
            </div>
          </Square>
        </div>
      </div>
    </Hold(p)>
  </div>
</Container>
`;
